watchtower/wtdb/range_index_test.go

```diff
@@ -0,0 +1,183 @@
+	require.EqualValues(t, 4, index.MaxHeight())
+	require.True(t, index.IsInIndex(3))
+
+	// Since the existing range ends at 4, test that adding the following
+	// numbers creates a new range.
+	index.Add(6)
+	index.Add(7)
+	index.Add(8)
+
+	require.EqualValues(t, 2, index.numRanges())
+	require.EqualValues(t, 4, index.getRangeMax(1))
+	require.EqualValues(t, 8, index.getRangeMax(6))
+	require.EqualValues(t, 8, index.MaxHeight())
+
+	index.Add(12)
+	index.Add(13)
+
+	require.EqualValues(t, 3, index.numRanges())
+	require.EqualValues(t, 4, index.getRangeMax(1))
+	require.EqualValues(t, 8, index.getRangeMax(6))
+	require.EqualValues(t, 13, index.getRangeMax(12))
+	require.EqualValues(t, 13, index.MaxHeight())
+
+	// Adding 5 should merge two of the ranges.
+	index.Add(5)
+
+	require.EqualValues(t, 2, index.numRanges())
+	require.EqualValues(t, 8, index.getRangeMax(1))
+	require.EqualValues(t, 13, index.getRangeMax(12))
+
+	index.Add(10)
+
+	require.EqualValues(t, 3, index.numRanges())
+	require.EqualValues(t, 8, index.getRangeMax(1))
+	require.EqualValues(t, 10, index.getRangeMax(10))
+	require.EqualValues(t, 13, index.getRangeMax(12))
+	require.EqualValues(t, 11, index.NumInSet())
+
+	index.Add(11)
+
+	require.EqualValues(t, 2, index.numRanges())
+	require.EqualValues(t, 8, index.getRangeMax(1))
+	require.EqualValues(t, 13, index.getRangeMax(10))
+
+	index.Add(9)
+
+	require.EqualValues(t, 1, index.numRanges())
+	require.EqualValues(t, 13, index.getRangeMax(1))
+	require.EqualValues(t, 13, index.MaxHeight())
+
+	// Make sure that things already in the range dont affect the index.
+	index.Add(9)
+	index.Add(7)
+	index.Add(2)
+	require.EqualValues(t, 1, index.numRanges())
+	require.EqualValues(t, 13, index.getRangeMax(1))
+	require.EqualValues(t, 13, index.MaxHeight())
+
+	// Test the transaction commit & revert.
+	tx := index.StartTx()
+	changes := tx.GetChanges(20)
+	require.NotNil(t, changes)
+	require.EqualValues(t, 20, *changes.NewIndexKey)
+	require.EqualValues(t, 20, *changes.NewIndexValue)
+	require.Nil(t, changes.DeleteIndex)
+
+	tx.Revert()
+	require.EqualValues(t, 1, index.numRanges())
+	require.EqualValues(t, 13, index.getRangeMax(1))
+	require.EqualValues(t, 13, index.MaxHeight())
+	require.True(t, index.IsInIndex(13))
+	require.False(t, index.IsInIndex(18))
+
+	tx = index.StartTx()
+	changes = tx.GetChanges(20)
+	require.NotNil(t, changes)
+	require.EqualValues(t, 20, *changes.NewIndexKey)
+	require.EqualValues(t, 20, *changes.NewIndexValue)
+	require.Nil(t, changes.DeleteIndex)
+	tx.Commit(changes)
+
+	require.EqualValues(t, 2, index.numRanges())
+	require.EqualValues(t, 13, index.getRangeMax(1))
+	require.EqualValues(t, 20, index.getRangeMax(20))
+	require.EqualValues(t, 20, index.MaxHeight())
+	require.EqualValues(t, 14, index.NumInSet())
+	require.True(t, index.IsInIndex(13))
+	require.False(t, index.IsInIndex(18))
+	require.False(t, index.IsInIndex(0))
+
+	// Test adding ranges to the set.
```

Comment:
Can likely use sub-tests `t.Run` to further visually/semantically divide the tests here.